tests/types/test_schema.py

```diff
@@ -285,13 +287,103 @@ def test_schema_inference_on_dictionary(dict_of_ndarrays):
         ]
     )
     # test exception is raised if non-numpy data in dictionary
-    match = "Data in the dictionary must be of type numpy.ndarray"
-    with pytest.raises(TypeError, match=match):
+    match = "Invalid values in dictionary. If passing a dictionary"
+    with pytest.raises(MlflowException, match=match):
         _infer_schema({"x": 1})
-    with pytest.raises(TypeError, match=match):
+    with pytest.raises(MlflowException, match=match):
         _infer_schema({"x": [1]})
 
 
+def test_schema_inference_on_string_input():
+    schema = _infer_schema("some string")
+    assert schema == Schema([ColSpec(DataType.string)])
+
+    with pytest.raises(TypeError, match="Expected one of the following types:"):
+        _infer_schema(1)
+
+
+@pytest.mark.parametrize(
+    "data, valid",
+    [
+        ({"a": "b", "c": "d"}, True),
+        ({"a": ["a", "b"], "b": ["c", "d"]}, True),
+        ({"a": "a", "b": ["a", "b"]}, True),
+        ({"a": 1, "b": "c"}, False),
+        ({"a": 1, "b": ["a", "b"]}, False),
+    ],
+)
+def test_schema_inference_on_dictionary_of_strings(data, valid):
+    if valid:
+        schema = _infer_schema(data)
+        assert schema == Schema([ColSpec(DataType.string, name) for name in data.keys()])
+    else:
+        with pytest.raises(
+            MlflowException, match="Invalid values in dictionary. If passing a dictionary"
+        ):
+            _infer_schema(data)
+
+
+@pytest.mark.parametrize(
+    "data, valid",
+    [
+        ({"a": "b", "b": "c"}, True),
+        ({1: "a", "b": "c"}, False),
+        ({12.4: "c", "d": "e"}, False),
+    ],
+)
+def test_schema_inference_validating_dictionary_keys(data, valid):
+    if valid:
+        schema = _infer_schema(data)
+        assert schema == Schema([ColSpec(DataType.string, name) for name in data.keys()])
+    else:
+        with pytest.raises(
+            MlflowException, match="The dictionary keys are not all strings. Invalid "
+        ):
+            _infer_schema(data)
```

Comment:
The same here: https://github.com/mlflow/mlflow/pull/8181/files#r1161372261

Reply:
Refactored all added tests to use a function-scoped for loop for valid and invalid examples